pkg/alertmanager/amcfg.go

```diff
@@ -0,0 +1,502 @@
+	i := 0
+	for k := range amConfigs {
+		amConfigIdentifiers[i] = k
+		i++
+	}
+	sort.Strings(amConfigIdentifiers)
+
+	subRoutes := []*route{}
+	for _, amConfigIdentifier := range amConfigIdentifiers {
+		crKey := types.NamespacedName{
+			Name:      amConfigs[amConfigIdentifier].Name,
+			Namespace: amConfigs[amConfigIdentifier].Namespace,
+		}
+
+		// add routes to subRoutes
+		if amConfigs[amConfigIdentifier].Spec.Route != nil {
+			subRoutes = append(subRoutes, convertRoute(amConfigs[amConfigIdentifier].Spec.Route, crKey, true))
+		}
+
+		// add receivers to baseConfig.Receivers
+		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
+			baseConfig.Receivers = append(baseConfig.Receivers, cg.convertReceiver(ctx, &receiver, crKey))
+		}
+
+		// add inhibitRules to baseConfig.InhibitRules
+		for _, inhibitRule := range amConfigs[amConfigIdentifier].Spec.InhibitRules {
+			baseConfig.InhibitRules = append(baseConfig.InhibitRules, convertInhibitRule(&inhibitRule))
+		}
+	}
+
+	// Append subroutes from base to the end, then replace with the new slice
+	subRoutes = append(subRoutes, baseConfig.Route.Routes...)
+	baseConfig.Route.Routes = subRoutes
+
+	return yaml.Marshal(baseConfig)
+}
+
+func convertRoute(in *monitoringv1alpha1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+
+	// Enforce continue to be true for main Route in a CR
+	cont := in.Continue
+	if firstLevelRoute {
+		cont = true
+	}
+
+	match := map[string]string{}
+	matchRE := map[string]string{}
+
+	for _, matcher := range in.Matchers {
+		if *matcher.Regex {
+			matchRE[matcher.Name] = matcher.Value
+		} else {
+			match[matcher.Name] = matcher.Value
+		}
+	}
+	if firstLevelRoute {
+		match["namespace"] = crKey.Namespace
+		delete(matchRE, "namespace")
+	}
+
+	if len(match) == 0 {
+		match = nil
+	}
+	if len(matchRE) == 0 {
+		matchRE = nil
+	}
+
+	var routes []*route = nil
+	if len(in.Routes) > 0 {
+		routes := make([]*route, len(in.Routes))
+		for i := range in.Routes {
+			routes[i] = convertRoute(&in.Routes[i], crKey, true)
+		}
+	}
+
+	receiver := prefixReceiverName(in.Receiver, crKey)
+
+	return &route{
+		Receiver:       receiver,
+		GroupByStr:     in.GroupBy,
+		GroupWait:      in.GroupWait,
+		GroupInterval:  in.GroupInterval,
+		RepeatInterval: in.RepeatInterval,
+		Continue:       cont,
+		Match:          match,
+		MatchRE:        matchRE,
+		Routes:         routes,
+	}
+}
+
+func (cg *configGenerator) convertReceiver(ctx context.Context, in *monitoringv1alpha1.Receiver, crKey types.NamespacedName) *receiver {
+
+	var pagerdutyConfigs []*pagerdutyConfig
+	if l := len(in.PagerDutyConfigs); l > 0 {
+		pagerdutyConfigs = make([]*pagerdutyConfig, l)
+		for i := range in.PagerDutyConfigs {
+			pagerdutyConfigs[i] = cg.convertPagerdutyConfig(ctx, in.PagerDutyConfigs[i], crKey)
+		}
+	}
+
+	return &receiver{
+		Name:             prefixReceiverName(in.Name, crKey),
+		PagerdutyConfigs: pagerdutyConfigs,
+	}
```

Comment:
I've left this one as-is for now -- as we add additional receivers, I don't think we'd want to return if there aren't any PagerDutyConfigs or EmailConfigs, we'd just look at the subsequent receiver types. I think it will become clearer as we add the other receiver types, and we can update it then as needed.

Reply:
:+1: